opencopilot/analytics.py

```
@@ -0,0 +1,155 @@
+    try:
+        pip_freeze_output = subprocess.check_output(["pip", "freeze"]).decode("utf-8")
+        matching_lines = [
+            l for l in pip_freeze_output.split("\n") if "opencopilot" in l
+        ]
+
+        for line in matching_lines:
+            if line.startswith(package_name):
+                # Normal PyPI install
+                declared_version = line.split("==")[1]
+                break
+            elif line.startswith(f"-e git+") and "egg=opencopilot_ai" in line:
+                # Local install from version-controlled repo
+                declared_version = line.split("-e ")[1]
+                break
+            elif "opencopilot-ai" in line or "opencopilot_ai" in line:
+                # Local install from non-version-controlled directory
+                declared_version = line
+    except CalledProcessError:
+        pass
+
+    return declared_version
+
+
+def hashed(s: str):
+    return xxhash.xxh64(s.encode("utf-8")).hexdigest()
+
+
+def get_hashed_user_id():
+    """Returns a hashed unique identifier for the user."""
+    mac_address: int = uuid.getnode()
+    return hashed(str(mac_address))
+
+
+def track(event_name: str, *args, **kwargs):
+    """Should be the entry point to all tracking."""
+    tracking_enabled = settings.get().TRACKING_ENABLED
+
+    if not tracking_enabled:
+        return
+
+    if event_name == "copilot_start":
+        _track_copilot_start(*args, **kwargs)
+    elif event_name == "copilot_start_error":
+        _track_copilot_start_error(*args, **kwargs)
+    elif event_name == "cli_command":
+        _track_cli_command(*args, **kwargs)
+    elif event_name == "cli_error":
+        _track_cli_error(*args, **kwargs)
+    elif event_name == "chat_message":
+        _track_chat_message(*args, **kwargs)
+    elif event_name == "api_error":
+        _track_api_error(*args, **kwargs)
+    else:
+        raise Exception(f"Unknown tracking event name: {event_name}")
```

Much better flow would be to either:
* use direct tracking functions, for example analytics.track_copilot_start()
* event_name as enum

With current approach we can face an embarrasing situation where we track an unkown tracking event and the whole app crashes because of it. In any case I think the analytics should raise an exception that can halt the whole App!

---

Using an enum now.